pool-apps/pool/src/lib/downstream/extensions_message_handler.rs

```
@@ -0,0 +1,130 @@
+        }
+
+        // Check which required extensions the client didn't request
+        let missing_required: Vec<u16> = required_extensions
+            .iter()
+            .filter(|ext| !requested.contains(ext))
+            .copied()
+            .collect();
+
+        // Determine response based on spec rules:
+        // - Success: If at least one extension is supported AND all required extensions are present
+        // - Error: If no extensions are supported OR required extensions are missing
+        let should_send_error = supported.is_empty() || !missing_required.is_empty();
+
+        if should_send_error {
+            // Send error response
+            error!(
+                "Downstream {}: Extension negotiation error: requested={:?}, supported={:?}, unsupported={:?}, missing_required={:?}",
+                self.downstream_id, requested, supported, unsupported, missing_required
+            );
+
+            let error = RequestExtensionsError {
+                request_id: msg.request_id,
+                unsupported_extensions: Seq064K::new(unsupported)
+                    .map_err(|_| PoolError::InvalidUnsupportedExtensionsSequence)?,
+                required_extensions: Seq064K::new(missing_required.clone())
+                    .map_err(|_| PoolError::InvalidRequiredExtensionsSequence)?,
+            };
+
+            let frame: Sv2Frame = AnyMessage::Extensions(error.into_static().into()).try_into()?;
+            self.downstream_channel
+                .downstream_sender
+                .send(frame)
+                .await?;
+
+            // If required extensions are missing, the server SHOULD disconnect the client
+            if !missing_required.is_empty() {
+                error!(
+                    "Downstream {}: Client does not support required extensions {:?}. Server MUST disconnect.",
+                    self.downstream_id, missing_required
+                );
+                // TODO: Disconnect the client
```

You can return an error, and the downstream gonna get disconnected